File: Online-judge/leetcode/lexicographically-smallest-equivalent-string.cpp
# include <bits/stdc++.h>
using namespace std;

/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/longest-path-with-different-adjacent-characters/description/
 * Time complexity  O((N+M)log∣Σ∣)
 * Space complexity O(∣Σ∣)
*/

class Solution {
private:
    array<int, 26> representative;
    char find(int x) {
        if (representative[x] == x) return representative[x];
        return representative[x] = find(representative[x]);
    }
    void makeRepresentative(int a , int b) {
        int x = find(a);
        int y = find(b);
        if (x == y) return;
        if (x < y) representative[y] = x;
        else representative[x] = y;
    }
public:
    string smallestEquivalentString(string s1, string s2, string baseStr) {
        string ans;
        for (int i = 0; i < 26; i++) {
            representative[i] = i;;
        }
        for (int index = 0; index < s1.size(); index++) {
            makeRepresentative(s1[index] - 'a', s2[index] - 'a');
        }
        for (char ch : baseStr) {
            ans += (char) find(ch - 'a') + 'a';
        }
        return ans;
    }
};

File: Online-judge/leetcode/lexicographically-smallest-equivalent-string.py
class Solution:
    representative = {}
    def find(self, ch):
        if self.representative[ch] == ch:
            return self.representative[ch]
        self.representative[ch] = self.find(self.representative[ch])
        return self.representative[ch]
    
    def makeRepresentative(self, x, y):
        xx = self.find(x)
        yy = self.find(y)
        if xx == yy:
            return
        if xx < yy:
            self.representative[yy] = xx
        else:
            self.representative[xx] = yy
        
    def smallestEquivalentString(self, s1: str, s2: str, baseStr: str) -> str:
        for i in range(ord('a'), ord('z')+1):
            self.representative[chr(i)] = chr(i)
        for index in range(len(s1)):
            self.makeRepresentative(s1[index], s2[index])
            
        ans = ''
        for index in range(len(baseStr)):
            ans += self.find(baseStr[index])
        return ans
